lib/common/acl.c

```diff
@@ -84,49 +84,42 @@ __xml_acl_create(xmlNode *xml, GList *acls, enum xml_private_flags mode)
                   crm_element_name(xml), acl->xpath);
 
     } else {
-        int offset = 0;
-        char buffer[MAX_XPATH_LEN];
+        GString *buffer = NULL;
 
         if (tag) {
-            offset += snprintf(buffer + offset, MAX_XPATH_LEN - offset,
-                               "//%s", tag);
+            buffer = pcmk__str_append(buffer, "//%s", tag);
         } else {
-            offset += snprintf(buffer + offset, MAX_XPATH_LEN - offset,
-                               "//*");
+            buffer = pcmk__str_append(buffer, "//*");
         }
 
         if (ref || attr) {
-            offset += snprintf(buffer + offset, MAX_XPATH_LEN - offset,
-                               "[");
+            buffer = pcmk__str_append(buffer, "[");
         }
 
         if (ref) {
-            offset += snprintf(buffer + offset, MAX_XPATH_LEN - offset,
-                               "@id='%s'", ref);
+            buffer = pcmk__str_append(buffer, "@id='%s'", ref);
         }
 
         // NOTE: schema currently does not allow this
         if (ref && attr) {
-            offset += snprintf(buffer + offset, MAX_XPATH_LEN - offset,
-                               " and ");
+            buffer = pcmk__str_append(buffer, " and ");
         }
 
         if (attr) {
-            offset += snprintf(buffer + offset, MAX_XPATH_LEN - offset,
-                               "@%s", attr);
+            buffer = pcmk__str_append(buffer, "@%s", attr);
         }
 
         if (ref || attr) {
-            offset += snprintf(buffer + offset, MAX_XPATH_LEN - offset,
-                               "]");
+            buffer = pcmk__str_append(buffer, "]");
         }
 
-        CRM_LOG_ASSERT(offset > 0);
-        acl->xpath = strdup(buffer);
+        CRM_LOG_ASSERT(buffer && buffer->len > 0);
+        acl->xpath = g_strdup(buffer->str);
```

Comment:
This is where things get tricky. C library memory management can't be mixed with glib memory management. Since acl->xpath will be freed by free(), it must be allocated by a standard function. I believe gchar is guaranteed to be an alias for char, so just casting it inside strdup() should be sufficient.

Ideally we wouldn't duplicate the memory at all, we'd just adopt buffer->str and call g_string_free(buffer, FALSE). Unfortunately that would mix the memory management, so we do need to strdup().

Reply:
> C library memory management can't be mixed with glib memory management

Ah, makes sense that the allocation needs to match the free. I wasn't immediately sure whether I should be using `g_strdup` just because the argument is a `gchar`.

> I believe gchar is guaranteed to be an alias for char, so just casting it inside strdup() should be sufficient.

It is, based on the API doc.

> so just casting it inside strdup() should be sufficient.

Can I rely on implicit casts for `gchar -> char`, or should I cast explicitly? Either would work, so seems like whatever style the project prefers.
Related: what about `gchar -> xmlChar`, which is defined as `unsigned char`?

> Ideally we wouldn't duplicate the memory at all, we'd just adopt buffer->str and call g_string_free(buffer, FALSE).

If we decide to make everything a `gchar/GString` in the future, then that sounds like the way to go. Is that the long-term goal if performance is up to par, or only switch to glib where needed?

Reply:
> Can I rely on implicit casts for `gchar -> char`, or should I cast explicitly? Either would work, so seems like whatever style the project prefers.

Not sure if any static analyzers will complain without the cast, but probably worth making explicit just in case. (Doesn't matter for style or correctness ...)

> Related: what about `gchar -> xmlChar`, which is defined as `unsigned char`?

A cast shouldn't cause any trouble. We have a convenience alias pcmkXmlStr for const xmlChar *

Reply:
> If we decide to make everything a `gchar/GString` in the future, then that sounds like the way to go. Is that the long-term goal if performance is up to par, or only switch to glib where needed?

Definitely not a goal, there is some memory/CPU overhead so I'd only use it where standard functions are incredibly inconvenient